api/envoy/type/matcher/metadata.proto

```diff
@@ -0,0 +1,126 @@
+//
+// .. code-block:: yaml
+//
+//    filter: envoy.filters.http.rbac
+//    path:
+//    - key: a
+//    - key: b
+//    - key: c
+//    values:
+//    - string_match:
+//        exact: ex
+//    - string_match:
+//        prefix: pr
+//
+// The following MetadataMatcher is not matched as the path [a, t] is pointing to a list value in
+// the Metadata which is not supported for now.
+//
+// .. code-block:: yaml
+//
+//    filter: envoy.filters.http.rbac
+//    path:
+//    - key: a
+//    - key: t
+//    values:
+//    - string_match:
+//        exact: m
+//
+// An example use of MetadataMatcher is specifying additional metadata in envoy.filters.http.rbac to
+// enforce access control based on dynamic metadata in a request. See :ref:`Permission
+// <envoy_api_msg_config.rbac.v2alpha.Permission>` and :ref:`Principal
+// <envoy_api_msg_config.rbac.v2alpha.Principal>`.
+message MetadataMatcher {
+  // Specifies the segment in a path to retrieve value from Metadata.
+  // Note: Currently it's not supported to retrieve a value from a list in Metadata. This means it
+  // will always be not matched if the associated value of the key is a list.
+  message PathSegment {
+    oneof segment {
+      option (validate.required) = true;
+
+      // If specified, use the key to retrieve the value in a Struct.
+      string key = 1 [(validate.rules).string.min_bytes = 1];
+    }
+  }
+
+  // Specifies the value to match. Only primitive value are supported. For non-primitive values, the
+  // result is always not matched.
+  message Value {
+    // NullMatch is an empty message to specify a null value.
+    message NullMatch {
+    }
+
+    // Specifies how to match a value.
+    oneof match_pattern {
+      option (validate.required) = true;
+
+      // If specified, a match occurs if and only if the target value is a NullValue.
+      NullMatch null_match = 1;
+
+      // If specified, a match occurs if and only if the target value is a double value and is
+      // matched to this field.
+      DoubleMatcher double_match = 2;
+
+      // If specified, a match occurs if and only if the target value is a string value and is
+      // matched to this field.
+      StringMatcher string_match = 3;
+
+      // If specified, a match occurs if and only if the target value is a bool value and is equal
+      // to this field.
+      bool bool_match = 4;
+
+      // If specified, value match will be performed based on whether the path is referring to a
+      // valid primitive value in the metadata. If the path is referring to a non-primitive value,
+      // the result is always not matched.
+      bool present_match = 5;
+    }
+  }
+
+  // The filter name to retrieve the Struct from the Metadata.
+  string filter = 1 [(validate.rules).string.min_bytes = 1];
+
+  // The path to retrieve the Value from the Struct.
+  repeated PathSegment path = 2 [(validate.rules).repeated .min_items = 1];
```

Comment:
Shall we use [fieldmask](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/field-mask)? You can write paths as "a.b.c".

Reply:
Fieldmask have some disadvantages, from my investigations, it cannot support either list or map which means we cannot use it directly here, unless we implement it by ourselves in Envoy. 

Second, the semantics would be much more complicated, it won't be `a.b.c`, you have to list all field names in the path, so it will be something like `fields.a.struct_value.fields.b.struct_value.fields.c.string_value`.

Last, Envoy already have a well implemented [metadataValue()](https://github.com/envoyproxy/envoy/blob/2bda6297e40605f806c76c75feb60bd1050f7cfe/source/common/config/metadata.cc#L6) method which satisfy all our needs here, so I think no need to use the fieldmask anymore.